File: ros2_ws/ser_cli/src/py_order_system/py_order_system/waiter_client.py
import sys
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

class WaiterClientAsync(Node):
    def __init__(self):
        super().__init__('waiter_node')
        # 1. 使用 create_client 建立客戶端 (服務型態 AddTwoInts, 服務名稱 'place_order')
        self.cli = self.create_client(AddTwoInts, 'place_order')
        
        # 2. 寫一個 while 迴圈，等待廚房上線 (wait_for_service)，若沒上線則持續印出「呼叫廚房中...」
        while not self.cli.wait_for_service(timeout_sec = 1.0):
            self.get_logger.info("呼叫廚房中...")
        
        # 3. 實例化 Request 物件 (AddTwoInts.Request())
        self.req = AddTwoInts.Request()

    def send_order(self, burgers, fries):
        # 4. 將參數 burgers, fries 填入 request 的 a 與 b
        self.req.a = burgers
        self.req.b = fries
        # 5. 使用 call_async 發送請求，並將結果存入 self.future
        self.future = self.cli.call_async(self.req)
        # 6. 使用 rclpy.spin_until_future_complete 等待廚房回應
        rclpy.spin_until_future_complete(self, self.future)
        # 7. 回傳 self.future.result()
        return self.future.result()

def main(args=None):
    # 8. 初始化 rclpy、建立 WaiterClientAsync 節點
    rclpy.init(args=args)
    waiter_client = WaiterClientAsync()
    # 9. 呼叫 send_order 送出 (2份漢堡, 3份薯條) 的請求
    waiter_client.get_logger().info('送出訂單: 漢堡 2 份, 薯條 3 份')
    response = waiter_client.send_order(2, 3)
    # 10. 印出結果 (response.sum)，然後銷毀節點與關閉系統
    waiter_client.get_logger().info(f'總數: {response.sum}')
    waiter_client.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws/ser_cli/src/cpp_order_system/src/waiter_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("waiter_node");
  
  // 建立客戶端，指定與 Server 相同的服務型態與名稱 ('place_order')
  auto client = node->create_client<example_interfaces::srv::AddTwoInts>("place_order");

  // 準備請求資料 (2份漢堡，3份薯條)
  auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
  request->a = 2;
  request->b = 3;

  // 使用 while 迴圈等待廚房上線，每秒檢查一次
  while (!client->wait_for_service(1s)) {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "呼叫廚房中，請稍後...");
  }

  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "服務生送出訂單：2 漢堡, 3 薯條...");
  
  // 非同步發送請求 (async_send_request 不會卡死主程式)
  auto result = client->async_send_request(request);
  
  // 等待 Server 回傳結果，若成功則印出
  if (rclcpp::spin_until_future_complete(node, result) == rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "廚房回報：總共 %ld 份餐點準備中！", result.get()->sum);
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "點餐失敗，廚房無回應");
  }

  rclcpp::shutdown();
  return 0;
}